sdk/azidentity/azure_cli_credential_client.go

```diff
@@ -0,0 +1,155 @@
+	// The command that Azure CLI would be run
+	command := "az account get-access-token --output json"
+	if scopes != nil {
+
+		// covert the scopes to a resource string
+		resource := c.scopeToResource(scopes)
+
+		// Validate the resource to make sure it doesn't include shell-meta characters since it gets sent as a command line argument to Azure CLI
+		isResourceMatch, error := regexp.MatchString("^[0-9a-zA-Z-.:/]+$", resource)
+		if error != nil {
+			return nil, error
+		}
+		if !isResourceMatch {
+			return nil, fmt.Errorf(invalidResourceError)
+		}
+
+		command = command + " --resource " + resource
+	}
+	// Execute Azure CLI command(az account get-access-token --output json --resource) to get a token to authenticate
+	out, errout, err := credentialClient.getAzureCLIAccessToken(command)
+
+	// Determining Azure CLI errors
+	if err != nil {
+		isLoginError := strings.HasPrefix(errout, AzNotLogIn)
+
+		// Is Azure CLI installed or not
+		isWinError := strings.HasPrefix(errout, WinAzureCLIError)
+		isOtherOsError, err := regexp.MatchString("az:(.*)not found", errout)
+		if err != nil {
+			return nil, err
+		} else if isWinError || isOtherOsError {
+			return nil, &AuthenticationFailedError{inner: errors.New(AzureCLINotInstalled)}
+		}
+
+		// Is user log in or not
+		if isLoginError {
+			return nil, &AuthenticationFailedError{inner: errors.New(AzNotLogIn)}
+		}
+
+		return nil, &AuthenticationFailedError{inner: errors.New(errout)}
+	}
+
+	return c.createAccessToken(out)
+}
+
+func (c *azureCLICredentialClient) getAzureCLIAccessToken(command string) ([]byte, string, error) {
+	var stderr bytes.Buffer
+	var stdout bytes.Buffer
+	var cmd *exec.Cmd
+	var shell string
+	var executePara string
+
+	if runtime.GOOS == "windows" {
+		shell = "cmd.exe"
+		executePara = "/c"
+	} else {
+		shell = "/bin/sh"
+		executePara = "-c"
+	}
+
+	// Run Azure CLI command to return resulting Access Token or error
+	cmd = exec.Command(shell, executePara, command)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+
+	return stdout.Bytes(), stderr.String(), err
+}
+
+func (c *azureCLICredentialClient) createAccessToken(output []byte) (*azcore.AccessToken, error) {
+	token := azureCLIAccessToken{}
+	accessToken := &azcore.AccessToken{}
+
+	err := json.Unmarshal(output, &token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse expiresOn string to date as required time format
+	dateString := token.ExpiresOn
+	timeformat := "2006-01-02 15:04:05.999999"
+	expiresOnValue, err := time.ParseInLocation(timeformat, dateString, time.Local)
+	if err != nil {
+		return nil, err
+	}
+
+	accessToken.ExpiresOn = expiresOnValue.In(time.UTC)
+	accessToken.Token = token.Token
+
+	return accessToken, err
+}
+
+func (c *azureCLICredentialClient) scopeToResource(scopes []string) string {
+	if !strings.HasSuffix(scopes[0], DefaultSuffix) {
```

Comment:
Some comments explaining what this method is trying to do would be useful. It looks like if the 0th scope doesn't end with a suffix, you return it. Why? What does this mean?

If it does end with that suffix, then you scan the string looking for ".default" (which is not in a const variable unlike DefaultSuffix) and returns a string up to that. But, what if ".default" is not in the string at all?

Reply:
The following code will remove the suffix from any scopes passed into the method since the Azure CLI command expect a resource string instead of a scope string. 
And some comments added to explaining what is this func purpose.